osquery/tables/system/posix/openssl_utils.cpp

```diff
@@ -9,226 +9,392 @@
-      // Generate the string representation of the issuer.
-      char* issuerBytes = X509_NAME_oneline(issuerName, nullptr, 0);
-      if (issuerBytes != nullptr) {
-        issuer = std::string(issuerBytes);
-        OPENSSL_free(issuerBytes);
-      }
+  std::string output;
+
+  for (const auto& key : kKeyUsageFlagList) {
+    const auto& usage_flag = key.first;
+    if ((key_usage_flags & usage_flag) == 0) {
+      continue;
     }
+
+    if (!output.empty()) {
+      output += ", ";
+    }
+
+    const auto& usage_name = key.second;
+    output += usage_name;
   }
 
-  X509_NAME* subjectName = X509_get_subject_name(cert);
-  if (subjectName == nullptr) {
-    return;
+  return output;
+}
+
+boost::optional<std::string> getCertificateSerialNumber(X509* cert) {
+  auto serial = X509_get_serialNumber(cert);
+  if (serial == nullptr) {
+    return boost::none;
   }
 
-  {
-    // Generate the string representation of the subject.
-    char* subjectBytes = X509_NAME_oneline(subjectName, nullptr, 0);
-    if (subjectBytes != nullptr) {
-      subject = std::string(subjectBytes);
-      OPENSSL_free(subjectBytes);
-    }
+  auto big_num_serial = ASN1_INTEGER_to_BN(serial, nullptr);
+  if (big_num_serial == nullptr) {
+    return boost::none;
   }
 
-  int nid = OBJ_txt2nid("CN");
+  char* string_serial = BN_bn2hex(big_num_serial);
+  OPENSSL_free(big_num_serial);
 
-  int index = X509_NAME_get_index_by_NID(subjectName, nid, -1);
-  if (index == -1) {
-    return;
+  if (string_serial == nullptr) {
+    return boost::none;
   }
 
-  X509_NAME_ENTRY* commonNameEntry = X509_NAME_get_entry(subjectName, index);
-  if (commonNameEntry == nullptr) {
-    return;
+  std::string output(string_serial);
+  OPENSSL_free(string_serial);
+
+  return output;
+}
+
+boost::optional<std::string> getCertificateAuthorityKeyID(X509* cert) {
+  auto cert_key_id = X509_get0_authority_key_id(cert);
+  if (cert_key_id == nullptr) {
+    return boost::none;
   }
 
-  ASN1_STRING* commonNameData = X509_NAME_ENTRY_get_data(commonNameEntry);
+  return convertBinaryDigestToString(cert_key_id->data, cert_key_id->length);
+}
 
-  const auto* data = ASN1_STRING_get0_data(commonNameData);
+boost::optional<std::string> getCertificateSubjectKeyID(X509* cert) {
+  auto cert_key_id = X509_get0_subject_key_id(cert);
+  if (cert_key_id == nullptr) {
+    return boost::none;
+  }
 
-  common_name = std::string(reinterpret_cast<const char*>(data));
+  return convertBinaryDigestToString(cert_key_id->data, cert_key_id->length);
 }
 
-std::string genKIDProperty(const unsigned char* data, int len) {
-  std::stringstream key_id;
-  for (int i = 0; i < len; i++) {
-    key_id << std::setw(2) << std::hex << std::setfill('0') << (int)data[i];
+boost::optional<std::string> getCertificateIssuerName(X509* cert) {
+  auto issuer_name = X509_get_issuer_name(cert);
+  if (issuer_name == nullptr) {
+    return boost::none;
+  }
+
+  auto string_issuer_name = X509_NAME_oneline(issuer_name, nullptr, 0);
```

Comment:
Reading [here](https://www.openssl.org/docs/man1.1.1/man3/X509_NAME_oneline.html) it seems that the use of this function is discouraged.

`X509_NAME_print_ex` is suggested as the alternative.

Reply:
This would be a breaking change, as it changes how the common name is displayed. I agree we should move away from deprecated functions, and I am thinking what would be the best way to provide users with a way good way to update their queries.

Would you be in favor of including a new hidden column for now with the new output, and mark the current `common_name` as deprecated? We could then do the switch in a future release. Tagging who already reviewed: @sharvilshah @directionless

#### OpenSSL documentation

```
The functions X509_NAME_oneline() and X509_NAME_print() produce a non standard
output form, they don't handle multi character fields and have various quirks
and inconsistencies. Their use is strongly discouraged in new applications
and they could be deprecated in a future release.
```


#### Original common_name, with X509_NAME_oneline

```
/C=US/ST=Arizona/L=Scottsdale/O=Starfield Technologies, Inc./CN=Starfield Services Root Certificate Authority - G2
```

#### New common_name, with X509_NAME_print_ex

```
C = US, ST = Arizona, L = Scottsdale, O = "Starfield Technologies, Inc.", CN = Starfield Services Root Certificate Authority - G2
```

Reply:
>Would you be in favor of including a new hidden column for now with the new output, and mark the current issuer as deprecated? We could then do the switch in a future release.

I think that would work.